unified-runtime/test/adapters/level_zero/enqueue_alloc.cpp

```diff
@@ -0,0 +1,374 @@
+                                        uur::GetDeviceUSMSingleSharedSupport},
+        EnqueueAllocMultiQueueTestParam{1024, 8, 256,
+                                        urEnqueueUSMDeviceAllocExp,
+                                        uur::GetDeviceUSMDeviceSupport},
+    }),
+    uur::deviceTestWithParamPrinter<EnqueueAllocMultiQueueTestParam>);
+
+TEST_P(urL0EnqueueAllocMultiQueueTest, Success) {
+  const size_t allocSize = std::get<1>(this->GetParam()).allocSize;
+  const size_t numQueues = std::get<1>(this->GetParam()).numQueues;
+  const size_t iterations = std::get<1>(this->GetParam()).iterations;
+  const auto enqueueUSMAllocFunc =
+      std::get<1>(this->GetParam()).funcParams.enqueueUSMAllocFunc;
+  const auto checkUSMSupportFunc =
+      std::get<1>(this->GetParam()).funcParams.checkUSMSupportFunc;
+
+  ur_device_usm_access_capability_flags_t USMSupport = 0;
+  ASSERT_SUCCESS(checkUSMSupportFunc(device, USMSupport));
+  if (!(USMSupport & UR_DEVICE_USM_ACCESS_CAPABILITY_FLAG_ACCESS)) {
+    GTEST_SKIP() << "Selected USM type is not supported.";
+  }
+
+  std::mutex mtx;
+  std::condition_variable cv1;
+  size_t cvCounter1 = numQueues;
+  std::condition_variable cv2;
+  size_t cvCounter2 = numQueues;
+
+  std::vector<std::thread> threads;
+  for (size_t idx = 0; idx < numQueues; idx++) {
+    // Create a thread per queue.
+    threads.emplace_back([&, idx, iterations] {
+      std::vector<void *> ptrs(iterations);
+      for (size_t i = 0; i < iterations; i++) {
+        ASSERT_SUCCESS(enqueueUSMAllocFunc(queues[idx], nullptr, allocSize,
+                                           nullptr, 0, nullptr, &ptrs[i],
+                                           nullptr));
+        ASSERT_NE(ptrs[i], nullptr);
+      }
+
+      ASSERT_SUCCESS(urQueueFinish(queues[idx]));
+
+      // Synchronize threads.
+      {
+        std::unique_lock<std::mutex> lock(mtx);
+        cvCounter1--;
+        if (cvCounter1 > 0) {
+          cv1.wait(lock, [&] { return cvCounter1 == 0; });
+        } else {
+          cv1.notify_all();
+        }
+      }
+
+      uint8_t fillPattern = 0xAF;
+      for (size_t i = 0; i < iterations; i++) {
+        ASSERT_SUCCESS(urEnqueueUSMFill(queues[idx], ptrs[i], 1, &fillPattern,
+                                        allocSize, 0, nullptr, nullptr));
+      }
+
+      ASSERT_SUCCESS(urQueueFinish(queues[idx]));
+
+      // Synchronize threads.
+      {
+        std::unique_lock<std::mutex> lock(mtx);
+        cvCounter2--;
+        if (cvCounter2 > 0) {
+          cv2.wait(lock, [&] { return cvCounter2 == 0; });
+        } else {
+          cv2.notify_all();
+        }
+      }
+
+      for (size_t i = 0; i < iterations; i++) {
+        ASSERT_SUCCESS(urEnqueueUSMFreeExp(queues[idx], nullptr, ptrs[i], 0,
+                                           nullptr, nullptr));
+      }
+
+      ASSERT_SUCCESS(urQueueFinish(queues[idx]));
+    });
+  }
+
+  for (auto &thread : threads) {
+    thread.join();
+  }
+}
+
```

Comment:
add a multi-queue test that checks that the buffers are actually reused a) between alloc/free from a single queue, and b) between queues.

Reply:
I'm not sure if that's what exactly you had in mind, I've added 2 `SuccessReuse` test cases for `urL0EnqueueAllocTest` and `urL0EnqueueAllocMultiQueueSameDeviceTest` fixtures.